api/app/Http/Controllers/AuthController.php

```diff
@@ -115,6 +129,81 @@ public function token(Request $request, UserRepository $userRepository)
+        return $this->responder->setTransformer($this->app->make(AuthTokenTransformer::class))->item($token);
+    }
+
+    /**
+     * Redirect the user to the Provider authentication page.
+     *
+     * @param  string     $provider
+     * @param  Socialite  $socialite
+     * @return Response
+     */
+    public function redirectToProvider($provider, Socialite $socialite)
+    {
+        $this->validateProvider($provider);
+
+        return $socialite->with($provider)->stateless()->redirect();
+    }
+
+    /**
+     * Obtain the user information from Provider.
+     *
+     * @param  string          $provider
+     * @param  Socialite       $socialite
+     * @param  UserRepository  $repository
+     * @return Response
+     */
+    public function handleProviderCallback($provider, Socialite $socialite, UserRepository $repository)
+    {
+        $this->validateProvider($provider);
+
+        $socialUser = $socialite->with($provider)->stateless()->user();
+
+        // Verify so we received an email address, if using oAuth credentials
+        // with Twitter for instance, that isn't whitelisted, no email
+        // address will be returned with the response.
+        // See the notes in Spira API doc under Social Login for more info.
+        if (!$socialUser->email) {
+            // The app is connected with the service, but the 3rd party service
+            // is not configured or allowed to return email addresses, so we
+            // can't process the data further. Let's throw an exception.
+            throw new UnprocessableEntityException('User object has no email');
+        }
+
+        // Parse the social user to fit within Spira's user model
+        $socialUser = ParserFactory::parse($socialUser, $provider);
+
+        // Get or create the Spira user from the social login
+        try {
+            $user = $repository->findByEmail($socialUser->email);
+        } catch (ModelNotFoundException $e) {
+            $user = $repository->getNewModel();
+            $user->fill(array_merge($socialUser->toArray(), ['user_type' => 'guest']));
+            $user = $repository->save($user);
+        }
+
+        $socialLogin = new SocialLogin(['provider' => $provider, 'token' => $socialUser->token]);
+        $user->addSocialLogin($socialLogin);
+
+        // Prepare response data
+        $token = $this->jwtAuth->fromUser($user, ['method' => $provider]);
+        $returnUrl = $socialite->with($provider)->stateless()->returnUrl();
+
+        return $this->responder->redirect($returnUrl, 302, ['token' => $token]);
+    }
+
+    /**
+     * Check so the provider exists.
+     *
+     * @param  string  $provider
+     * @return void
+     */
+    protected function validateProvider($provider)
+    {
+        if (!in_array($provider, array_keys($this->app['config']['services']))) {
+            // Throws a NotFoundHttpException
+            $this->app->abort(404, 'Invalid provider');
```

Comment:
It is preferable to throw actual exceptions as it is clearer to the developer what is happening, and that a function might have a catchable exception (as would be indicated in the docblock).

This exception should actually be a `NotImplementedException` as theoretically we might implement more